cpp/src/cider/tests/nextgen/compiler/ArrayTest.cpp

```diff
@@ -0,0 +1,171 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include <google/protobuf/util/json_util.h>
+#include <gtest/gtest.h>
+#include <cstddef>
+#include <cstdint>
+#include <vector>
+
+#include "exec/nextgen/Nextgen.h"
+#include "exec/nextgen/context/Batch.h"
+#include "exec/plan/parser/SubstraitToRelAlgExecutionUnit.h"
+#include "exec/plan/parser/TypeUtils.h"
+
+#include "tests/TestHelpers.h"
+#include "tests/utils/ArrowArrayBuilder.h"
+#include "tests/utils/Utils.h"
+#include "util/Logger.h"
+
+using namespace cider::exec::nextgen;
+
+static const std::shared_ptr<CiderAllocator> allocator =
+    std::make_shared<CiderDefaultAllocator>();
+
+operators::TranslatorPtr initSqlToTranslators(const std::string& sql,
+                                              const std::string& create_ddl) {
+  // SQL Parsing
+  auto json = RunIsthmus::processSql(sql, create_ddl);
+  ::substrait::Plan plan;
+  google::protobuf::util::JsonStringToMessage(json, &plan);
+
+  generator::SubstraitToRelAlgExecutionUnit substrait2eu(plan);
+  auto eu = substrait2eu.createRelAlgExecutionUnit();
+
+  // Pipeline Building
+  auto pipeline = parsers::toOpPipeline(eu);
+  transformer::Transformer transformer;
+  return transformer.toTranslator(pipeline);
+}
+
+context::RuntimeCtxPtr executeAndReturnRuntimeCtx(const std::string& create_ddl,
+                                                  const std::string& sql,
+                                                  ArrowArray* array) {
+  auto translators = initSqlToTranslators(sql, create_ddl);
+
+  // Codegen
+  context::CodegenContext codegen_ctx;
+  auto module = cider::jitlib::LLVMJITModule("test", true);
+  cider::jitlib::JITFunctionPointer function =
+      cider::jitlib::JITFunctionBuilder()
+          .registerModule(module)
+          .setFuncName("query_func")
+          .addReturn(cider::jitlib::JITTypeTag::VOID)
+          .addParameter(cider::jitlib::JITTypeTag::POINTER,
+                        "context",
+                        cider::jitlib::JITTypeTag::INT8)
+          .addParameter(cider::jitlib::JITTypeTag::POINTER,
+                        "input",
+                        cider::jitlib::JITTypeTag::INT8)
+          .addProcedureBuilder(
+              [&codegen_ctx, &translators](cider::jitlib::JITFunctionPointer func) {
+                codegen_ctx.setJITFunction(func);
+                translators->consume(codegen_ctx);
+                func->createReturn();
+              })
+          .build();
+  module.finish();
+  auto query_func = function->getFunctionPointer<void, int8_t*, int8_t*>();
+
+  // Execution
+  auto runtime_ctx = codegen_ctx.generateRuntimeCTX(allocator);
+  query_func((int8_t*)runtime_ctx.get(), (int8_t*)array);
+
+  return runtime_ctx;
+}
+
+class CiderArrayTest : public ::testing::Test {
+ public:
+  template <typename COL_TYPE>
```

Comment:
need to be template function? or we should use template type rather than int8_t in function body?